### PR TITLE
Add a minimum score option to the command line interface

### PR DESCRIPTION
**Pull request by RobertoSchneiders**
This is the first step for #98. It work, but, I'm not completely happy with some choices, maybe you can provide some nice suggestions :)

First of all, I'm assuming that the responsibility for the exit code should be in the cli, not in the lib itself. Also, the minimum score feature belongs to the cli and not the lib itself. So, I think it is plausible to expect the lib commands to return the score. The problem with that is not all commands will have a score (eg. --help), so, it's not a great solution. Anyway, I implemented that way for now. Any suggestions are welcome.

I made a change in the CI command, now it inherit from the default command and overrides the `critique` method (The only difference between them). I made this to remove some code duplication. I don't know if this was the best thing to do, so, a feedback will be appreciated.

### CODE REVIEW

**Comment by RobertoSchneiders** [comment]
Outputs:
```
$rubycritic lib --minimum-score 98
...
New critique at file:///...
Score (97.875) is below the minimum 98.0

$echo $?
1
```
```
rubycritic lib --minimum-score 97
...
New critique at file:///...
Score: 97.875

$echo $?
0
```

File: lib/rubycritic/cli/application.rb [hunk 1]
```diff
@@ -14,12 +14,29 @@ def initialize(argv)
...
         STATUS_SUCCESS
       rescue OptionParser::InvalidOption => error
         $stderr.puts "Error: #{error}"
         STATUS_ERROR
       end
+
+      private
+
+      def satisfy_minimum_score?(result, options)
+        return true unless result.is_a?(Hash) && result[:score]
```
Comment (troessner) [inline lib/rubycritic/cli/application.rb]: Let's not do this. This is 1.) extremely obscure (no innocent reader will understand why you are doing this at first glance) and 2.) makes heavy use of type checking, something that is smelly per se in a dynamically typed language like Ruby.
We should test here that we satisfy the minimum-score, nothing else. The check that this option is set should happen above, in "execute".

File: lib/rubycritic/cli/application.rb [hunk 2]
```diff
@@ -14,12 +14,29 @@ def initialize(argv)
...
+
+        status = result[:score] >= options.minimum_score
+        if status
+          $stdout.puts "Score: #{result[:score]}"
+        else
+          $stderr.puts "Score (#{result[:score]}) is below the minimum #{options.minimum_score}"
+        end
```
Comment (troessner) [inline lib/rubycritic/cli/application.rb]: The method is a predicate method. Problem is, that you're doing quite a few other things in here. Callers of this method would expect a boolean value back, but not multiple side effects like printing to stdout and so on.

File: lib/rubycritic/cli/application.rb [hunk 3]
```diff
@@ -14,12 +14,29 @@ def initialize(argv)
 
       def execute
         parsed_options = @options.parse
-        ::Rubycritic::CommandFactory.create(parsed_options).execute
+        result = ::Rubycritic::CommandFactory.create(parsed_options).execute
```
Comment (troessner) [inline lib/rubycritic/cli/application.rb]: I believe "execute" should rather return a kind of Reporter initialized with the given options from the options_parser. And this reporter you can then ask: 

* "Is my score to low?" (with the reporter knowing if the flag has been set or not)
* "What should I return?"

You could check out how we do it in [`Reek`](https://github.com/troessner/reek/blob/master/lib/reek/cli/application.rb#L31) - as you can see, the "Application" still has access to the corresponding exit code even at the very end of the life time.

I'm thinking about something along the lines (Pseudo code obviously):

```Ruby
reporter = ::Rubycritic::CommandFactory.create(parsed_options).execute
reporter.exit_status
```

With the reporter knowing the options and then encapsulating this logic.

Comment (troessner) [inline lib/rubycritic/cli/application.rb]: And since we're on it, let's drop the scope operator as well. Just

>> Rubycritic::CommandFactory.create...

is fine.

File: lib/rubycritic/cli/application.rb (hunk 1 above)
Comment (troessner) [inline lib/rubycritic/cli/application.rb]: Addendum: In reference to my comment above, this whole logic should not be here but in a Reporter.
Without any type check of course ;)

File: lib/rubycritic/cli/application.rb (hunk 2 above)
Comment (troessner) [inline lib/rubycritic/cli/application.rb]: Or in one sentence: This should actually be multiple extremely small methods.

**Comment by troessner** [comment]
A general comment: Of course every pull request should have extensive test coverage - I assume you just didn't add any unit tests because you considered this as a "feeling out" process.
However, it would help to add 2 integration tests right from the start: 1 that tests our exit_status without "--minimum-score" and one that tests our exit_status with it. 
As much as I dislike cucumber, this would be very fitting here. We don't have it integrated at the moment, but I think this would be a good start. You can alway check out how we do it in [Reek](https://github.com/troessner/reek/tree/master/features/command_line_interface)

**Comment by RobertoSchneiders** [comment]
Thanks for the help @troessner. 

I never worked with cucumber, so, it is the first time I configure and used it, but, I tried to do more or less the same of reek. I hope it is close to what it should be. 

I created the integration tests, I'm not sure if they are good enough, but it is a start. :)

About the reporter. I looked at the reek code and probably I don't completely understood your line of thinking. Rubicritic already have a reporter, but it is for json and html. Even so, I implemented the way I understood your idea, but, I had some problems. I decided to push the code anyway and start a discussion.

Problems with the actual implementation:
* I had to pass the full options hash to all commands
* I had to use inheritance to avoid a lot of duplication code in the commands (Don't know if worth it)
* Rubocop is complaining about a method that I'm using for inhertance override, which breaked the build (I'm looking for a solution)

I'm not english native, so, probably I made some typos, I will review this tomorrow.

Thanks for your patience.

File: features/command_line_interface/exit_status.feature [hunk 4]
```diff
@@ -0,0 +1,20 @@
+Feature: Return the correct exit status codes
+  In order to use Rubycritic in Continuous Integration environments
+  Rubycritic returns the exit status code properly
+
+  Scenario: Status indicates a success when analyse a clean file
+    Given the clean file 'clean.rb'
+    When I run rubycritic clean.rb
+    Then the exit status indicates a success
```
Comment (troessner) [inline features/command_line_interface/exit_status.feature]: Quick style thing: Let's use newlines between scenarios.

File: features/command_line_interface/exit_status.feature [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  Scenario: Status indicates a success when analyse a smelly file
+    Given the smelly file 'smelly.rb'
+    When I run rubycritic smelly.rb
+    Then the exit status indicates a success
+  Scenario: Status indicates a success when analyse a clean file with the minimum score option
+    Given the clean file 'clean.rb'
+    When I run rubycritic --minimum-score 100 clean.rb
+    Then the exit status indicates a success
+  Scenario: Status indicates an error when analyse a smelly file with the minimum score option
+    Given the smelly file 'smelly.rb'
```
Comment (troessner) [inline features/command_line_interface/exit_status.feature]: Can we make this step more expressive by actually telling the caller what score this file has explicitly?
Thinking of something like

>> Given the smelly file 'smelly.rb' with a score of 115

File: features/command_line_interface/minimum_score.feature [hunk 6]
```diff
@@ -0,0 +1,30 @@
+Feature: Brake if overall score is bellow minimum
+  In order to breake the Continuous Integration builds based on a score threshold
+  Rubycritic returns the exit status according with the score
+
+  Scenario: Status indicates an error when score below the minimum
+    Given the smelly file 'smelly.rb'
+    When I run rubycritic --minimum-score 100 smelly.rb
+    Then the exit status indicates an error
+  Scenario: Status indicates a success when score is above the minimum
+    Given the smelly file 'smelly.rb'
```
Comment (troessner) [inline features/command_line_interface/minimum_score.feature]: Same here - I, as the user have **no** idea what the real score is. So that we're using "50", "100" and so on is confusing to me. Let's put all of those numbers in the step definition so that it's obvious.

File: features/command_line_interface/exit_status.feature [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    When I run rubycritic --minimum-score 100 smelly.rb
+    Then the exit status indicates an error
```
Comment (troessner) [inline features/command_line_interface/exit_status.feature]: I should have read the whole PR first ;) - this is an duplication to features/command_line_interface/minimum_score.feature (which you probably are aware of). I can see why you would do that but I still think it's redundant. I'd rather scrap this file completely and just use features/command_line_interface/minimum_score.feature. That you then can use this for CI is kind of implicitly.

File: features/command_line_interface/minimum_score.feature [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    """
+    Score: 100
+    """
+  Scenario: Prints a message informing the score is below the minimum
+    Given the empty file 'empty.rb'
+    When I run rubycritic --minimum-score 100 empty.rb
+    Then the output should contain:
+    """
+    Score (0.0) is below the minimum 100.0
+    """
```
Comment (troessner) [inline features/command_line_interface/minimum_score.feature]: Nice feature, that's exactly what I had in mind!

File: lib/rubycritic/cli/application.rb [hunk 9]
```diff
@@ -9,17 +9,24 @@ class Application
       STATUS_ERROR   = 1
 
       def initialize(argv)
+        puts argv.class
```
Comment (troessner) [inline lib/rubycritic/cli/application.rb]: I'm pretty you don't want this here ;)

File: lib/rubycritic/cli/options.rb [hunk 10]
```diff
@@ -27,6 +27,10 @@ def parse
             @format = format
           end
 
+          opts.on("-s", "--minimum-score [MIN_SCORE]", "Set a minimum score") do |min_score|
+            @minimum_score = min_score.to_f
```
Comment (troessner) [inline lib/rubycritic/cli/options.rb]: Don't we want to use an integer here?
And let's rather use proper constructor methods:

>>
[1] pry(main)> nil.to_i
=> 0
[2] pry(main)> Integer(nil)
TypeError: can't convert nil into Integer
from (pry):2:in `Integer'

File: lib/rubycritic/command_factory.rb [hunk 11]
```diff
@@ -8,16 +8,16 @@ def self.create(options = {})
       case Config.mode
       when :version
         require "rubycritic/commands/version"
-        Command::Version.new
+        Command::Version.new(options_hash)
```
Comment (troessner) [inline lib/rubycritic/command_factory.rb]: Let's not use [hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation) - I think just calling it "options" is fine here.

File: lib/rubycritic/commands/status_reporter.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      SUCCESS = 0
+      SCORE_BELOW_MINIMUM = 1
+
+      def initialize(options)
+        @options = options
+        @status = SUCCESS
+      end
+
+      def score=(score)
+        return if score.nil?
```
Comment (troessner) [inline lib/rubycritic/commands/status_reporter.rb]: Please see my comment [here](https://github.com/whitesmith/rubycritic/pull/99/files#diff-c017e99a0be9d85fac1d8d58af3382a2R31)

File: lib/rubycritic/commands/status_reporter.rb [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+        @score = score.to_f
+        update_status
+        update_status_message
```
Comment (troessner) [inline lib/rubycritic/commands/status_reporter.rb]: That's confusing since it's not obvious what the difference between those 2 methods.
Additionally, you're decoupling something that **must** be coupled, right? There's no point in updating the status without updating the message?
I'd rather merge "update_status_message" with "update_status"

**Comment by troessner** [comment]
Overall: Very nice! This is going in the right direction!
Let's address all comments and then I'm confident we can get this one merged middle of next week ;)

File: lib/rubycritic/cli/application.rb (hunk 9 above)
Comment (RobertoSchneiders) [inline lib/rubycritic/cli/application.rb]: Ooops!! :grimacing:

File: features/command_line_interface/exit_status.feature (hunk 5 above)
Comment (RobertoSchneiders) [inline features/command_line_interface/exit_status.feature]: I opted for doing this because I was afraid that the score for the smelly code can change overtime, and this test will break frequently. That's not the case or this is not a problem?

File: lib/rubycritic/command_factory.rb (hunk 11 above)
Comment (RobertoSchneiders) [inline lib/rubycritic/command_factory.rb]: I haven't named that variable, but I agree, I will fix this. :+1:

File: lib/rubycritic/commands/status_reporter.rb (hunk 12 above)
Comment (RobertoSchneiders) [inline lib/rubycritic/commands/status_reporter.rb]: Nice catch.

File: lib/rubycritic/cli/application.rb (hunk 9 above)
Comment (troessner) [inline lib/rubycritic/cli/application.rb]: A wiser man than me once gave me a tip I still follow up until this day: Never indent your debugging statements. That way they'll always stick out like a sore thumb ;)

File: features/command_line_interface/exit_status.feature (hunk 5 above)
Comment (troessner) [inline features/command_line_interface/exit_status.feature]: >> That's not the case or this is not a problem?

Actually I think this is good because that you actually realise that something has changed.
Check out [this feature](https://github.com/troessner/reek/blob/master/features/samples.feature) in `Reek`. 
It is huge. We could  easily make that a lot thinner and remove some duplication. We don't. Because if we change something deep down and do NOT expect anything to change in `Reek`s behaviour and all of a sudden we have some changes there we know something went south.
Also, I'd take readability over churn any day, so let's go for readability ;)

**Comment by RobertoSchneiders** [comment]
@troessner I made some changes. 

At some point I breaked the help functionality and I only noticed that by lucky in a manual test, so, I created a feature test for it. :)

**Comment by RobertoSchneiders** [comment]
A test is breaking only on Travis, I'm trying to figure it out..

**Comment by RobertoSchneiders** [comment]
It was breaking because of a change on the master branch. Fixed.

File: features/command_line_interface/minimum_score.feature [hunk 14]
```diff
@@ -0,0 +1,39 @@
+Feature: Brake if overall score is bellow minimum
```
Comment (Onumis) [inline features/command_line_interface/minimum_score.feature]: typo: `Break`

File: features/command_line_interface/options.feature [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+                                             html (default)
+                                             json
+                                             console
+          -s, --minimum-score [MIN_SCORE]  Set a minimum score
+          -m, --mode-ci                    Use CI mode (faster, but only analyses last commit)
+              --deduplicate-symlinks       De-duplicate symlinks based on their final target
+              --suppress-ratings           Suppress letter ratings
+          -v, --version                    Show gem's version
+          -h, --help                       Show this message
+
```
Comment (troessner) [inline features/command_line_interface/options.feature]: Very nice!

File: lib/rubycritic/command_factory.rb [hunk 16]
```diff
@@ -3,21 +3,24 @@
...
-        Command::Help.new(options.help_text)
+        Command::Help
       when :ci
         require "rubycritic/commands/ci"
-        Command::Ci.new(options_hash[:paths])
+        Command::Ci
       else
         require "rubycritic/commands/default"
-        Command::Default.new(options_hash[:paths])
+        Command::Default
```
Comment (troessner) [inline lib/rubycritic/command_factory.rb]: Nice refactoring!

File: lib/rubycritic/commands/status_reporter.rb [hunk 17]
```diff
@@ -0,0 +1,43 @@
+module Rubycritic
+  module Command
+    class StatusReporter
+      attr_reader :status, :status_message
+      SUCCESS = 0
+      SCORE_BELOW_MINIMUM = 1
```
Comment (troessner) [inline lib/rubycritic/commands/status_reporter.rb]: Lets use those constants in our features above instead of the magic numbers  0 and 1 directly.

**Comment by troessner** [comment]
Great job! One last comment from my side and then lets get this one merged!

File: lib/rubycritic/commands/status_reporter.rb (hunk 17 above)
Comment (RobertoSchneiders) [inline lib/rubycritic/commands/status_reporter.rb]: :+1:

**Comment by troessner** [comment]
Almost there! Now let's go the last step: Please squash your pull request into  one single commit and then I'll happily merge this one!

**Comment by RobertoSchneiders** [comment]
@troessner Done :)

File: .travis.yml [hunk 18]
```diff
@@ -13,4 +13,5 @@ matrix:
     - rvm: jruby
   fast_finish: true
 
+before_install: "gem update bundler"
```
Comment (troessner) [inline .travis.yml]: Wait, why this change here? I fail to see how this belongs to this branch?

File: lib/rubycritic/commands/base.rb [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+module Rubycritic
+  module Command
+    class Base
+      def initialize(options)
+        @options = options
+        @status_reporter = Rubycritic::Command::StatusReporter.new(@options)
+      end
+
+      def execute
+        @status_reporter
```
Comment (troessner) [inline lib/rubycritic/commands/base.rb]: I had an uneasy feeling about this in the first review already and on second thought I don't think we should do this:

You're using the template pattern here where you overwrite `execute` in subclasses. The problem with the way do it right now is that you're creating an undocumented, obscure dependency here - corresponding methods in subclasses **must** call `super`, because otherwise the return type changes signifcantly. And not only that, they need to call `super` in a certain order as well.

Let's not do this, this will confuse ourselves in 4 weeks from now. How about just raising NotImplementedError here and then return "@status_report" in all `executes` below?

File: .travis.yml (hunk 18 above)
Comment (troessner) [inline .travis.yml]: Ah, I see where this is coming from. Pardon my french, but you fucked up the rebase.
This change and a **lot** of other changes in this PR are coming from the one [with the ConsoleReport](https://github.com/whitesmith/rubycritic/pull/91) we recently merged.
I'm afraid you'll have to remove all changes that slipped into your rebase before we can merge this.
The joys of git ;)

File: lib/rubycritic/core/smell.rb [hunk 20]
```diff
@@ -8,7 +8,7 @@ class Smell
 
     attribute :context
     attribute :cost
-    attribute :locations
+    attribute :locations, Array, :default => []
```
Comment (troessner) [inline lib/rubycritic/core/smell.rb]: Comes from https://github.com/whitesmith/rubycritic/pull/91

File: lib/rubycritic/generators/text/list.rb [hunk 21]
```diff
@@ -0,0 +1,43 @@
+require "colorize"
```
Comment (troessner) [inline lib/rubycritic/generators/text/list.rb]: Changes below come from [here](https://github.com/whitesmith/rubycritic/pull/91) as well.

**Comment by troessner** [comment]
I'm afraid this is not ready to merge. 2 things:

1.) My comment in regards to lib/rubycritic/commands/base.rb - that should be easy to fix though.
2.) The second problem is more severe: You seemed to have mixed up your code with [the ConsoleReport PR](https://github.com/whitesmith/rubycritic/pull/91) - we need to remove all the code that came in from here.

**Comment by RobertoSchneiders** [comment]
Sorry about the rebase @troessner , I forgot to remove that commit. 

I made the refactor about the execute method.

**Comment by Onumis** [comment]
Looks good to me :+1:

File: .rubocop.yml [hunk 22]
```diff
@@ -517,6 +518,7 @@ Style/TrivialAccessors:
     - to_str
     - to_s
     - to_sym
+    - execute
```
Comment (RobertoSchneiders) [inline .rubocop.yml]: I removed this, but for some reason the diff is not updated..

**Comment by troessner** [comment]
There seems to be an ongoing fight between you and git - looking at the current diff, we moved one step forward but two back ;)

1.) The `execute` method is back to calling `super` - looks like your latest changes are lost
2.) Like before: we need to squash this to one commit, right now its 18

**Comment by RobertoSchneiders** [comment]
embarassing... I pushed the wrong branch. Sorry again! Fixed! 

Build failed, but, failed in `Updating installed gems`, seems to be something wrong with bundler. I can't restart the builds.

**Comment by troessner** [comment]
Yes, seems like rubygems is the problem here. I restarted the build twice already :(
Let's retry tomorrow.

**Comment by troessner** [comment]
Travis is green, merging this one. Great job, thanks!

**Comment by RobertoSchneiders** [comment]
Thanks for your help (and for the patience) @troessner, sorry for the git mess. It was a great experience, learned a lot.
